Crash During File Watcher Worker Restart

### What happened?

Occasionally and without consistent patterns, our Laravel app—which runs with FrankenPHP and Octane—crashes when a file is saved. To provide some context, we switched from the default Laravel Octane watcher, which is based on Chokidar, to the new FrankenPHP e-dant file watcher that was recently merged. We made this move due to frequent crashes with the previous setup that occurred without logs or any apparent means to reproduce the issue.

Our experience with the new e-dant based file watcher has been excellent so far; it is much faster and more reliable. However, we still experience rare crashes. but, at least we now receive logs when crashes occur.

![image](https://github.com/user-attachments/assets/4d696388-57f2-4bbd-b47d-9a978d988948)

I've been working with the team to attempt to reliably reproduce the crash at all, but it seems to happen without any consistent pattern of use.

I can however give you folks our exact `Dockerfile`
```dockerfile
FROM dunglas/frankenphp-dev:php8.3-bookworm

# Update and Install apt packages
RUN apt update \
  && apt install -y --no-install-recommends \
    libgmp-dev \
  && apt clean && rm -rf /var/lib/apt/lists/*

# Adding needed Laravel extensions
# https://laravel.com/docs/deployment#server-requirements
# a lot of these are pre-installed, and will skip
RUN install-php-extensions \
  ctype \
  curl \
  dom \
  fileinfo \
  filter \
  hash \
  intl \
  mbstring \
  openssl \
  pcre \
  session \
  tokenizer \
  xml \
  # Adding Laravel Octane required extensions
  pcntl \
  # Adding Laravel Dusk required extensions
  zip \
  # Adding XRDM required extensions
  pdo_pgsql \
  redis \
  opcache \
  # Adding OTEL required extensions
  opentelemetry \
  ## Needed for parquet file parser
  gmp \
  bcmath

ENV OPCACHE_REVALIDATE_FREQ=0
ENV OPCACHE_VALIDATE_TIMESTAMPS=0
ENV OPCACHE_MAX_ACCELERATED_FILES=20000
ENV OPCACHE_MAX_WASTED_PERCENTAGE=10
ENV OPCACHE_MEMORY_CONSUMPTION=192
ENV OPCACHE_INTERNED_STRINGS_BUFFER=16
ENV OPCACHE_ENABLE_FILE_OVERRIDE=1
ENV OPCACHE_JIT=tracing
ENV OPCACHE_JIT_BUFFER_SIZE=0

COPY php.ini $PHP_INI_DIR/php.ini
ENV PHP_MEMORY_LIMIT=512M

EXPOSE 80

COPY Caddyfile /etc/caddy/Caddyfile
ENV CADDYFILE=/etc/caddy/Caddyfile

ENV HOST=0.0.0.0
ENV PORT=80

ENV WORKERS=auto
ENV MAX_REQUEST=500

ENV ADMIN_HOST=localhost
ENV ADMIN_PORT=2019

WORKDIR /app

CMD [ "/bin/sh", "-c", "php artisan octane:frankenphp --caddyfile=$CADDYFILE --host=$HOST --port=$PORT --workers=$WORKERS --max-requests=$MAX_REQUEST --admin-host=$ADMIN_HOST --admin-port=$ADMIN_PORT ${HTTPS:+--https} ${HTTP_REDIRECT:+--http-redirect} ${WATCH:+--watch}"]
```

Our `php.ini`
```ini
; Default PHP directives https://www.php.net/manual/en/ini.list.php

[PHP]
short_open_tag = Off

expose_php = Off

display_errors = Off
display_startup_errors = E_ALL & ~E_DEPRECATED & ~E_STRICT
log_errors = On

max_execution_time = ${PHP_MAX_EXECUTION_TIME}
memory_limit = ${PHP_MEMORY_LIMIT}

output_buffering = 4096
register_argc_argv = Off
request_order = "GP"
variables_order = "GPCS"

[opcache]
opcache.enable=1
opcache.enable_cli=1
opcache.use_cwd=0

opcache.revalidate_freq=${OPCACHE_REVALIDATE_FREQ}
opcache.validate_timestamps=${OPCACHE_VALIDATE_TIMESTAMPS}

opcache.max_accelerated_files=${OPCACHE_MAX_ACCELERATED_FILES}
opcache.max_wasted_percentage=${OPCACHE_MAX_WASTED_PERCENTAGE}
opcache.memory_consumption=${OPCACHE_MEMORY_CONSUMPTION}
opcache.interned_strings_buffer=${OPCACHE_INTERNED_STRINGS_BUFFER}
opcache.enable_file_override=${OPCACHE_ENABLE_FILE_OVERRIDE}

opcache.jit=${OPCACHE_JIT}
opcache.jit_buffer_size=${OPCACHE_JIT_BUFFER_SIZE}
```

And our `Caddyfile` (based off of the [Laravel Octane stubbed Caddyfile](https://github.com/laravel/octane/blob/2.x/src/Commands/stubs/Caddyfile) with small changes)
```
{
	{$CADDY_GLOBAL_OPTIONS}

	servers {
		metrics
	}

	admin {$CADDY_SERVER_ADMIN_HOST}:{$CADDY_SERVER_ADMIN_PORT}

	frankenphp {
		worker {
			file "{$APP_PUBLIC_PATH}/frankenphp-worker.php" {$CADDY_SERVER_WORKER_COUNT}
			watch "{$APP_BASE_PATH}/**/*.php"
		}
	}
}

{$CADDY_SERVER_SERVER_NAME} {
	log {
		level {$CADDY_SERVER_LOG_LEVEL}

		# Redact the authorization query parameter that can be set by Mercure...
		format filter {
			wrap {$CADDY_SERVER_LOGGER}
			fields {
				uri query {
					replace authorization REDACTED
				}
			}
		}
	}

	route {
		root * "{$APP_PUBLIC_PATH}"
		encode zstd br gzip

		# Mercure configuration is injected here...
		{$CADDY_SERVER_EXTRA_DIRECTIVES}

		php_server {
			index frankenphp-worker.php
			# Required for the public/storage/ directory...
			resolve_root_symlink
		}
	}
}
```

Other potentially relevant info, is that run our Docker containers with Orbstack, and our docker compose looks something like this.

```yaml
services:
 web:
    build: Dockerfile # as seen above
    labels:
      - dev.orbstack.domains=some-url.local
    environment:
      WORKERS: 4 # settings workers to 4, as opposed to 'auto' for more consistency. Hopefully this isn't the issue 🤔 
    volumes:
      - ./laravel-code:/app/
```


### Build Type

Docker (Debian Bookworm)

### Worker Mode

Yes

### Operating System

GNU/Linux

### CPU Architecture

Apple Silicon

### PHP configuration

```shell
PHP Version => 8.3.12

System => Linux 6ac6f317cc77 6.10.12-orbstack-00282-gd1783374c25e #26 SMP Wed Oct  2 10:16:55 UTC 2024 aarch64
Build Date => Sep 27 2024 06:07:53
Build System => Linux - Docker
Build Provider => https://github.com/docker-library/php
Configure Command =>  './configure'  '--build=aarch64-linux-gnu' '--with-config-file-path=/usr/local/etc/php' '--with-config-file-scan-dir=/usr/local/etc/php/conf.d' '--enable-option-checking=fatal' '--with-mhash' '--with-pic' '--enable-mbstring' '--enable-mysqlnd' '--with-password-argon2' '--with-sodium=shared' '--with-pdo-sqlite=/usr' '--with-sqlite3=/usr' '--with-curl' '--with-iconv' '--with-openssl' '--with-readline' '--with-zlib' '--enable-phpdbg' '--enable-phpdbg-readline' '--with-pear' '--with-libdir=lib/aarch64-linux-gnu' '--enable-embed' '--enable-zts' '--disable-zend-signals' 'build_alias=aarch64-linux-gnu'
Server API => Command Line Interface
Virtual Directory Support => enabled
Configuration File (php.ini) Path => /usr/local/etc/php
Loaded Configuration File => /usr/local/etc/php/php.ini
Scan this dir for additional .ini files => /usr/local/etc/php/conf.d
Additional .ini files parsed => /usr/local/etc/php/conf.d/docker-php-ext-bcmath.ini,
/usr/local/etc/php/conf.d/docker-php-ext-gmp.ini,
/usr/local/etc/php/conf.d/docker-php-ext-intl.ini,
/usr/local/etc/php/conf.d/docker-php-ext-opcache.ini,
/usr/local/etc/php/conf.d/docker-php-ext-opentelemetry.ini,
/usr/local/etc/php/conf.d/docker-php-ext-pcntl.ini,
/usr/local/etc/php/conf.d/docker-php-ext-pcov.ini,
/usr/local/etc/php/conf.d/docker-php-ext-pdo_pgsql.ini,
/usr/local/etc/php/conf.d/docker-php-ext-redis.ini,
/usr/local/etc/php/conf.d/docker-php-ext-sodium.ini,
/usr/local/etc/php/conf.d/docker-php-ext-zip.ini

PHP API => 20230831
PHP Extension => 20230831
Zend Extension => 420230831
Zend Extension Build => API420230831,TS
PHP Extension Build => API20230831,TS
Debug Build => no
Thread Safety => enabled
Thread API => POSIX Threads
Zend Signal Handling => disabled
Zend Memory Manager => enabled
Zend Multibyte Support => provided by mbstring
Zend Max Execution Timers => enabled
IPv6 Support => enabled
DTrace Support => disabled

Registered PHP Streams => https, ftps, compress.zlib, php, file, glob, data, http, ftp, phar, zip
Registered Stream Socket Transports => tcp, udp, unix, udg, ssl, tls, tlsv1.0, tlsv1.1, tlsv1.2, tlsv1.3
Registered Stream Filters => zlib.*, convert.iconv.*, string.rot13, string.toupper, string.tolower, convert.*, consumed, dechunk

This program makes use of the Zend Scripting Language Engine:
Zend Engine v4.3.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.3.12, Copyright (c), by Zend Technologies


 _______________________________________________________________________


Configuration

bcmath

BCMath support => enabled

Directive => Local Value => Master Value
bcmath.scale => 0 => 0

Core

PHP Version => 8.3.12

Directive => Local Value => Master Value
allow_url_fopen => On => On
allow_url_include => Off => Off
arg_separator.input => & => &
arg_separator.output => & => &
auto_append_file => no value => no value
auto_globals_jit => On => On
auto_prepend_file => no value => no value
browscap => no value => no value
default_charset => UTF-8 => UTF-8
default_mimetype => text/html => text/html
disable_classes => no value => no value
disable_functions => no value => no value
display_errors => Off => Off
display_startup_errors => On => On
doc_root => no value => no value
docref_ext => no value => no value
docref_root => no value => no value
enable_dl => On => On
enable_post_data_reading => On => On
error_append_string => no value => no value
error_log => no value => no value
error_log_mode => 0644 => 0644
error_prepend_string => no value => no value
error_reporting => -1 => no value
expose_php => Off => Off
extension_dir => /usr/local/lib/php/extensions/no-debug-zts-20230831 => /usr/local/lib/php/extensions/no-debug-zts-20230831
fiber.stack_size => no value => no value
file_uploads => On => On
hard_timeout => 2 => 2
highlight.comment => <font style="color: #FF8000">#FF8000</font> => <font style="color: #FF8000">#FF8000</font>
highlight.default => <font style="color: #0000BB">#0000BB</font> => <font style="color: #0000BB">#0000BB</font>
highlight.html => <font style="color: #000000">#000000</font> => <font style="color: #000000">#000000</font>
highlight.keyword => <font style="color: #007700">#007700</font> => <font style="color: #007700">#007700</font>
highlight.string => <font style="color: #DD0000">#DD0000</font> => <font style="color: #DD0000">#DD0000</font>
html_errors => Off => Off
ignore_repeated_errors => Off => Off
ignore_repeated_source => Off => Off
ignore_user_abort => Off => Off
implicit_flush => On => On
include_path => .:/usr/local/lib/php => .:/usr/local/lib/php
input_encoding => no value => no value
internal_encoding => no value => no value
log_errors => On => On
mail.add_x_header => Off => Off
mail.force_extra_parameters => no value => no value
mail.log => no value => no value
mail.mixed_lf_and_crlf => Off => Off
max_execution_time => 0 => 0
max_file_uploads => 20 => 20
max_input_nesting_level => 64 => 64
max_input_time => -1 => -1
max_input_vars => 1000 => 1000
max_multipart_body_parts => -1 => -1
memory_limit => 512M => 512M
open_basedir => no value => no value
output_buffering => 0 => 0
output_encoding => no value => no value
output_handler => no value => no value
post_max_size => 8M => 8M
precision => 14 => 14
realpath_cache_size => 4096K => 4096K
realpath_cache_ttl => 120 => 120
register_argc_argv => On => On
report_memleaks => On => On
report_zend_debug => Off => Off
request_order => GP => GP
sendmail_from => no value => no value
sendmail_path => /usr/sbin/sendmail -t -i => /usr/sbin/sendmail -t -i
serialize_precision => -1 => -1
short_open_tag => Off => Off
SMTP => localhost => localhost
smtp_port => 25 => 25
sys_temp_dir => no value => no value
syslog.facility => LOG_USER => LOG_USER
syslog.filter => no-ctrl => no-ctrl
syslog.ident => php => php
unserialize_callback_func => no value => no value
upload_max_filesize => 2M => 2M
upload_tmp_dir => no value => no value
user_dir => no value => no value
user_ini.cache_ttl => 300 => 300
user_ini.filename => .user.ini => .user.ini
variables_order => GPCS => GPCS
xmlrpc_error_number => 0 => 0
xmlrpc_errors => Off => Off
zend.assertions => 1 => 1
zend.detect_unicode => On => On
zend.enable_gc => On => On
zend.exception_ignore_args => Off => Off
zend.exception_string_param_max_len => 15 => 15
zend.max_allowed_stack_size => 0 => 0
zend.multibyte => Off => Off
zend.reserved_stack_size => 0 => 0
zend.script_encoding => no value => no value

ctype

ctype functions => enabled

curl

cURL support => enabled
cURL Information => 7.88.1
Age => 10
Features
AsynchDNS => Yes
CharConv => No
Debug => No
GSS-Negotiate => No
IDN => Yes
IPv6 => Yes
krb4 => No
Largefile => Yes
libz => Yes
NTLM => Yes
NTLMWB => Yes
SPNEGO => Yes
SSL => Yes
SSPI => No
TLS-SRP => Yes
HTTP2 => Yes
GSSAPI => Yes
KERBEROS5 => Yes
UNIX_SOCKETS => Yes
PSL => Yes
HTTPS_PROXY => Yes
MULTI_SSL => No
BROTLI => Yes
ALTSVC => Yes
HTTP3 => No
UNICODE => No
ZSTD => Yes
HSTS => Yes
GSASL => No
Protocols => dict, file, ftp, ftps, gopher, gophers, http, https, imap, imaps, ldap, ldaps, mqtt, pop3, pop3s, rtmp, rtmpe, rtmps, rtmpt, rtmpte, rtmpts, rtsp, scp, sftp, smb, smbs, smtp, smtps, telnet, tftp
Host => aarch64-unknown-linux-gnu
SSL Version => OpenSSL/3.0.14
ZLib Version => 1.2.13
libSSH Version => libssh2/1.10.0

Directive => Local Value => Master Value
curl.cainfo => no value => no value

date

date/time support => enabled
timelib version => 2022.10
"Olson" Timezone Database Version => 2024.2
Timezone Database => internal
Default timezone => UTC

Directive => Local Value => Master Value
date.default_latitude => 31.7667 => 31.7667
date.default_longitude => 35.2333 => 35.2333
date.sunrise_zenith => 90.833333 => 90.833333
date.sunset_zenith => 90.833333 => 90.833333
date.timezone => UTC => UTC

dom

DOM/XML => enabled
DOM/XML API Version => 20031129
libxml Version => 2.9.14
HTML Support => enabled
XPath Support => enabled
XPointer Support => enabled
Schema Support => enabled
RelaxNG Support => enabled

fileinfo

fileinfo support => enabled
libmagic => 543

filter

Input Validation and Filtering => enabled

Directive => Local Value => Master Value
filter.default => unsafe_raw => unsafe_raw
filter.default_flags => no value => no value

gmp

gmp support => enabled
GMP version => 6.2.1

hash

hash support => enabled
Hashing Engines => md2 md4 md5 sha1 sha224 sha256 sha384 sha512/224 sha512/256 sha512 sha3-224 sha3-256 sha3-384 sha3-512 ripemd128 ripemd160 ripemd256 ripemd320 whirlpool tiger128,3 tiger160,3 tiger192,3 tiger128,4 tiger160,4 tiger192,4 snefru snefru256 gost gost-crypto adler32 crc32 crc32b crc32c fnv132 fnv1a32 fnv164 fnv1a64 joaat murmur3a murmur3c murmur3f xxh32 xxh64 xxh3 xxh128 haval128,3 haval160,3 haval192,3 haval224,3 haval256,3 haval128,4 haval160,4 haval192,4 haval224,4 haval256,4 haval128,5 haval160,5 haval192,5 haval224,5 haval256,5

MHASH support => Enabled
MHASH API Version => Emulated Support

iconv

iconv support => enabled
iconv implementation => glibc
iconv library version => 2.36

Directive => Local Value => Master Value
iconv.input_encoding => no value => no value
iconv.internal_encoding => no value => no value
iconv.output_encoding => no value => no value

intl

Internationalization support => enabled
ICU version => 72.1
ICU Data version => 72.1
ICU TZData version => 2022e
ICU Unicode version => 15.0

Directive => Local Value => Master Value
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => Off => Off

json

json support => enabled

libxml

libXML support => active
libXML Compiled Version => 2.9.14
libXML Loaded Version => 20914
libXML streams => enabled

mbstring

Multibyte Support => enabled
Multibyte string engine => libmbfl
HTTP input encoding translation => disabled
libmbfl version => 1.3.2

mbstring extension makes use of "streamable kanji code filter and converter", which is distributed under the GNU Lesser General Public License version 2.1.

Multibyte (japanese) regex support => enabled
Multibyte regex (oniguruma) version => 6.9.8

Directive => Local Value => Master Value
mbstring.detect_order => no value => no value
mbstring.encoding_translation => Off => Off
mbstring.http_input => no value => no value
mbstring.http_output => no value => no value
mbstring.http_output_conv_mimetypes => ^(text/|application/xhtml\+xml) => ^(text/|application/xhtml\+xml)
mbstring.internal_encoding => no value => no value
mbstring.language => neutral => neutral
mbstring.regex_retry_limit => 1000000 => 1000000
mbstring.regex_stack_limit => 100000 => 100000
mbstring.strict_detection => Off => Off
mbstring.substitute_character => no value => no value

mysqlnd

mysqlnd => enabled
Version => mysqlnd 8.3.12
Compression => supported
core SSL => supported
extended SSL => supported
Command buffer size => 4096
Read buffer size => 32768
Read timeout => 86400
Collecting statistics => Yes
Collecting memory statistics => No
Tracing => n/a
Loaded plugins => mysqlnd,debug_trace,auth_plugin_mysql_native_password,auth_plugin_mysql_clear_password,auth_plugin_caching_sha2_password,auth_plugin_sha256_password
API Extensions =>

openssl

OpenSSL support => enabled
OpenSSL Library Version => OpenSSL 3.0.14 4 Jun 2024
OpenSSL Header Version => OpenSSL 3.0.14 4 Jun 2024
Openssl default config => /usr/lib/ssl/openssl.cnf

Directive => Local Value => Master Value
openssl.cafile => no value => no value
openssl.capath => no value => no value

opentelemetry

opentelemetry hooks => enabled
extension version => 1.1.0

Directive => Local Value => Master Value
opentelemetry.allow_stack_extension => Off => Off
opentelemetry.attr_hooks_enabled => Off => Off
opentelemetry.attr_post_handler_function => OpenTelemetry\API\Instrumentation\WithSpanHandler::post => OpenTelemetry\API\Instrumentation\WithSpanHandler::post
opentelemetry.attr_pre_handler_function => OpenTelemetry\API\Instrumentation\WithSpanHandler::pre => OpenTelemetry\API\Instrumentation\WithSpanHandler::pre
opentelemetry.conflicts => no value => no value
opentelemetry.validate_hook_functions => On => On

pcntl

pcntl support => enabled

pcov

PCOV support => Enabled
PCOV version => 1.0.11
pcov.directory => /app/
pcov.exclude => none
pcov.initial.memory => 65336 bytes
pcov.initial.files => 64

pcre

PCRE (Perl Compatible Regular Expressions) Support => enabled
PCRE Library Version => 10.42 2022-12-12
PCRE Unicode Version => 14.0.0
PCRE JIT Support => enabled
PCRE JIT Target => ARM-64 64bit (little endian + unaligned)

Directive => Local Value => Master Value
pcre.backtrack_limit => 1000000 => 1000000
pcre.jit => On => On
pcre.recursion_limit => 100000 => 100000

PDO

PDO support => enabled
PDO drivers => sqlite, pgsql

pdo_pgsql

PDO Driver for PostgreSQL => enabled
PostgreSQL(libpq) Version => 15.8

pdo_sqlite

PDO Driver for SQLite 3.x => enabled
SQLite Library => 3.40.1

Phar

Phar: PHP Archive support => enabled
Phar API version => 1.1.1
Phar-based phar archives => enabled
Tar-based phar archives => enabled
ZIP-based phar archives => enabled
gzip compression => enabled
bzip2 compression => disabled (install ext/bz2)
Native OpenSSL support => enabled


Phar based on pear/PHP_Archive, original concept by Davey Shafik.
Phar fully realized by Gregory Beaver and Marcus Boerger.
Portions of tar implementation Copyright (c) 2003-2009 Tim Kientzle.
Directive => Local Value => Master Value
phar.cache_list => no value => no value
phar.readonly => On => On
phar.require_hash => On => On

posix

POSIX support => enabled

random

Version => 8.3.12

readline

Readline Support => enabled
Readline library => 8.2

Directive => Local Value => Master Value
cli.pager => no value => no value
cli.prompt => \b \>  => \b \>

redis

Redis Support => enabled
Redis Version => 6.1.0
Redis Sentinel Version => 1.0
Available serializers => php, json
Available compression => lzf, zstd, lz4

Directive => Local Value => Master Value
redis.arrays.algorithm => no value => no value
redis.arrays.auth => no value => no value
redis.arrays.autorehash => 0 => 0
redis.arrays.connecttimeout => 0 => 0
redis.arrays.consistent => 0 => 0
redis.arrays.distributor => no value => no value
redis.arrays.functions => no value => no value
redis.arrays.hosts => no value => no value
redis.arrays.index => 0 => 0
redis.arrays.lazyconnect => 0 => 0
redis.arrays.names => no value => no value
redis.arrays.pconnect => 0 => 0
redis.arrays.previous => no value => no value
redis.arrays.readtimeout => 0 => 0
redis.arrays.retryinterval => 0 => 0
redis.clusters.auth => no value => no value
redis.clusters.cache_slots => 0 => 0
redis.clusters.persistent => 0 => 0
redis.clusters.read_timeout => 0 => 0
redis.clusters.seeds => no value => no value
redis.clusters.timeout => 0 => 0
redis.pconnect.connection_limit => 0 => 0
redis.pconnect.echo_check_liveness => 1 => 1
redis.pconnect.pool_detect_dirty => 0 => 0
redis.pconnect.pool_pattern => no value => no value
redis.pconnect.pool_poll_timeout => 0 => 0
redis.pconnect.pooling_enabled => 1 => 1
redis.session.compression => none => none
redis.session.compression_level => 3 => 3
redis.session.early_refresh => 0 => 0
redis.session.lock_expire => 0 => 0
redis.session.lock_retries => 100 => 100
redis.session.lock_wait_time => 20000 => 20000
redis.session.locking_enabled => 0 => 0

Reflection

Reflection => enabled

session

Session Support => enabled
Registered save handlers => files user redis rediscluster
Registered serializer handlers => php_serialize php php_binary

Directive => Local Value => Master Value
session.auto_start => Off => Off
session.cache_expire => 180 => 180
session.cache_limiter => nocache => nocache
session.cookie_domain => no value => no value
session.cookie_httponly => Off => Off
session.cookie_lifetime => 0 => 0
session.cookie_path => / => /
session.cookie_samesite => no value => no value
session.cookie_secure => Off => Off
session.gc_divisor => 100 => 100
session.gc_maxlifetime => 1440 => 1440
session.gc_probability => 1 => 1
session.lazy_write => On => On
session.name => PHPSESSID => PHPSESSID
session.referer_check => no value => no value
session.save_handler => files => files
session.save_path => no value => no value
session.serialize_handler => php => php
session.sid_bits_per_character => 4 => 4
session.sid_length => 32 => 32
session.upload_progress.cleanup => On => On
session.upload_progress.enabled => On => On
session.upload_progress.freq => 1% => 1%
session.upload_progress.min_freq => 1 => 1
session.upload_progress.name => PHP_SESSION_UPLOAD_PROGRESS => PHP_SESSION_UPLOAD_PROGRESS
session.upload_progress.prefix => upload_progress_ => upload_progress_
session.use_cookies => On => On
session.use_only_cookies => On => On
session.use_strict_mode => Off => Off
session.use_trans_sid => Off => Off

SimpleXML

SimpleXML support => enabled
Schema support => enabled

sodium

sodium support => enabled
libsodium headers version => 1.0.18
libsodium library version => 1.0.18

SPL

SPL support => enabled
Interfaces => OuterIterator, RecursiveIterator, SeekableIterator, SplObserver, SplSubject
Classes => AppendIterator, ArrayIterator, ArrayObject, BadFunctionCallException, BadMethodCallException, CachingIterator, CallbackFilterIterator, DirectoryIterator, DomainException, EmptyIterator, FilesystemIterator, FilterIterator, GlobIterator, InfiniteIterator, InvalidArgumentException, IteratorIterator, LengthException, LimitIterator, LogicException, MultipleIterator, NoRewindIterator, OutOfBoundsException, OutOfRangeException, OverflowException, ParentIterator, RangeException, RecursiveArrayIterator, RecursiveCachingIterator, RecursiveCallbackFilterIterator, RecursiveDirectoryIterator, RecursiveFilterIterator, RecursiveIteratorIterator, RecursiveRegexIterator, RecursiveTreeIterator, RegexIterator, RuntimeException, SplDoublyLinkedList, SplFileInfo, SplFileObject, SplFixedArray, SplHeap, SplMinHeap, SplMaxHeap, SplObjectStorage, SplPriorityQueue, SplQueue, SplStack, SplTempFileObject, UnderflowException, UnexpectedValueException

sqlite3

SQLite3 support => enabled
SQLite Library => 3.40.1

Directive => Local Value => Master Value
sqlite3.defensive => On => On
sqlite3.extension_dir => no value => no value

standard

Dynamic Library Support => enabled
Path to sendmail => /usr/sbin/sendmail -t -i

Directive => Local Value => Master Value
assert.active => On => On
assert.bail => Off => Off
assert.callback => no value => no value
assert.exception => On => On
assert.warning => On => On
auto_detect_line_endings => Off => Off
default_socket_timeout => 60 => 60
from => no value => no value
session.trans_sid_hosts => no value => no value
session.trans_sid_tags => a=href,area=href,frame=src,form= => a=href,area=href,frame=src,form=
unserialize_max_depth => 4096 => 4096
url_rewriter.hosts => no value => no value
url_rewriter.tags => form= => form=
user_agent => no value => no value

tokenizer

Tokenizer Support => enabled

xml

XML Support => active
XML Namespace Support => active
libxml2 Version => 2.9.14

xmlreader

XMLReader => enabled

xmlwriter

XMLWriter => enabled

Zend OPcache

Opcode Caching => Up and Running
Optimization => Enabled
SHM Cache => Enabled
File Cache => Disabled
JIT => Disabled
Startup => OK
Shared memory model => mmap
Cache hits => 0
Cache misses => 2067
Used memory => 60995600
Free memory => 140330992
Wasted memory => 0
Interned Strings Used memory => 12962184
Interned Strings Free memory => 3815032
Cached scripts => 2067
Cached keys => 4057
Max keys => 32531
OOM restarts => 0
Hash keys restarts => 0
Manual restarts => 0
Start time => 2024-10-11T18:04:30+0000
Last restart time => none
Last force restart time => none

Directive => Local Value => Master Value
opcache.blacklist_filename => no value => no value
opcache.dups_fix => Off => Off
opcache.enable => On => On
opcache.enable_cli => On => On
opcache.enable_file_override => On => On
opcache.error_log => no value => no value
opcache.file_cache => no value => no value
opcache.file_cache_consistency_checks => On => On
opcache.file_cache_only => Off => Off
opcache.file_update_protection => 2 => 2
opcache.force_restart_timeout => 180 => 180
opcache.huge_code_pages => Off => Off
opcache.interned_strings_buffer => 16 => 16
opcache.jit => tracing => tracing
opcache.jit_bisect_limit => 0 => 0
opcache.jit_blacklist_root_trace => 16 => 16
opcache.jit_blacklist_side_trace => 8 => 8
opcache.jit_buffer_size => 0 => 0
opcache.jit_debug => 0 => 0
opcache.jit_hot_func => 127 => 127
opcache.jit_hot_loop => 64 => 64
opcache.jit_hot_return => 8 => 8
opcache.jit_hot_side_exit => 8 => 8
opcache.jit_max_exit_counters => 8192 => 8192
opcache.jit_max_loop_unrolls => 8 => 8
opcache.jit_max_polymorphic_calls => 2 => 2
opcache.jit_max_recursive_calls => 2 => 2
opcache.jit_max_recursive_returns => 2 => 2
opcache.jit_max_root_traces => 1024 => 1024
opcache.jit_max_side_traces => 128 => 128
opcache.jit_max_trace_length => 1024 => 1024
opcache.jit_prof_threshold => 0.005 => 0.005
opcache.lockfile_path => /tmp => /tmp
opcache.log_verbosity_level => 1 => 1
opcache.max_accelerated_files => 20000 => 20000
opcache.max_file_size => 0 => 0
opcache.max_wasted_percentage => 10 => 10
opcache.memory_consumption => 192 => 192
opcache.opt_debug_level => 0 => 0
opcache.optimization_level => 0x7FFEBFFF => 0x7FFEBFFF
opcache.preferred_memory_model => no value => no value
opcache.preload => no value => no value
opcache.preload_user => no value => no value
opcache.protect_memory => Off => Off
opcache.record_warnings => Off => Off
opcache.restrict_api => no value => no value
opcache.revalidate_freq => 0 => 0
opcache.revalidate_path => Off => Off
opcache.save_comments => On => On
opcache.use_cwd => Off => Off
opcache.validate_permission => Off => Off
opcache.validate_root => Off => Off
opcache.validate_timestamps => Off => Off

zip

Zip => enabled
Zip version => 1.22.3
Libzip version => 1.7.3
BZIP2 compression => Yes
XZ compression => No
ZSTD compression => No
AES-128 encryption => Yes
AES-192 encryption => Yes
AES-256 encryption => Yes

zlib

ZLib Support => enabled
Stream Wrapper => compress.zlib://
Stream Filter => zlib.inflate, zlib.deflate
Compiled Version => 1.2.13
Linked Version => 1.2.13

Directive => Local Value => Master Value
zlib.output_compression => Off => Off
zlib.output_compression_level => -1 => -1
zlib.output_handler => no value => no value
```


### Relevant log output

```shell
panic: workers "public/frankenphp-worker.php": too many consecutive failures.  
   INFO  goroutine 2986 [running]:  
   INFO  github.com/dunglas/frankenphp.(*worker).startNewWorkerThread.func2().  
   INFO  /go/srcworker.go:165 +0xe4.  
   INFO  sync.(*Once).doSlow(0x4000606a20?, 0x1957320?).  
   INFO  /usr/local/go/src/sync/once.go:74 +0x100.  
   INFO  sync.(*Once).Do(...).  
   INFO  /usr/local/go/src/sync/once.go:65.  
   INFO  github.com/dunglas/frankenphp.(*worker).startNewWorkerThread(0x40006e2cf0).  
   INFO  /go/srcworker.go:159 +0x428.  
   INFO  created by github.com/dunglas/frankenphp.initWorkers in goroutine 13.  
   INFO  /go/srcworker.go:51 +0x164.
```

--- a/testdata/failing-worker.php
+++ b/testdata/failing-worker.php
@@ -1,9 +1,9 @@
 <?php
 
-$fail = random_int(1, 100) < 1;
-$wait = random_int(1, 5);
+$fail = random_int(1, 100) < 10;
+$wait = random_int(1000 * 100, 1000 * 500); // wait 100ms - 500ms
 
-sleep($wait);
+usleep($wait);
 if ($fail) {
     exit(1);
 }

--- a/worker.go
+++ b/worker.go
@@ -25,16 +25,16 @@ type worker struct {
 
 const maxWorkerErrorBackoff = 1 * time.Second
 const minWorkerErrorBackoff = 100 * time.Millisecond
-const maxWorkerConsecutiveFailures = 10
+const maxWorkerConsecutiveFailures = 5
 
 var (
+    isInitialBoot    bool = true
 	workersReadyWG   sync.WaitGroup
 	workerShutdownWG sync.WaitGroup
 	workersAreReady  atomic.Bool
 	workersAreDone   atomic.Bool
 	workersDone      chan interface{}
 	workers          map[string]*worker = make(map[string]*worker)
-	watcherIsEnabled bool
 )
 
 func initWorkers(opt []workerOpt) error {
@@ -55,6 +55,7 @@ func initWorkers(opt []workerOpt) error {
 
 	workersReadyWG.Wait()
 	workersAreReady.Store(true)
+	isInitialBoot = false
 
 	return nil
 }
@@ -163,13 +164,12 @@ func (worker *worker) startNewWorkerThread() {
 			// if we end up here, the worker has not been up for backoff*2
 			// this is probably due to a syntax error or another fatal error
 			if failureCount >= maxWorkerConsecutiveFailures {
-				if !watcherIsEnabled {
+				if isInitialBoot {
 					panic(fmt.Errorf("workers %q: too many consecutive failures", worker.fileName))
 				}
-				logger.Warn("many consecutive worker failures", zap.String("worker", worker.fileName))
-			} else {
-				failureCount += 1
+				logger.Warn("many consecutive worker failures", zap.String("worker", worker.fileName), zap.Int("failures", failureCount))
 			}
+			failureCount += 1
 		})
 		backingOffLock.RLock()
 		wait := backoff
@@ -210,7 +210,6 @@ func restartWorkersOnFileChanges(workerOpts []workerOpt) error {
 	if len(directoriesToWatch) == 0 {
 		return nil
 	}
-	watcherIsEnabled = true
 	restartWorkers := func() {
 		restartWorkers(workerOpts)
 	}

--- a/worker.go
+++ b/worker.go
@@ -28,7 +28,7 @@ const minWorkerErrorBackoff = 100 * time.Millisecond
 const maxWorkerConsecutiveFailures = 5
 
 var (
-    isInitialBoot    bool = true
+	isInitialBoot    bool = true
 	workersReadyWG   sync.WaitGroup
 	workerShutdownWG sync.WaitGroup
 	workersAreReady  atomic.Bool

--- a/worker.go
+++ b/worker.go
@@ -25,7 +25,7 @@ type worker struct {
 
 const maxWorkerErrorBackoff = 1 * time.Second
 const minWorkerErrorBackoff = 100 * time.Millisecond
-const maxWorkerConsecutiveFailures = 60
+const maxWorkerConsecutiveFailures = 10
 
 var (
 	workersReadyWG   sync.WaitGroup
@@ -34,6 +34,7 @@ var (
 	workersAreDone   atomic.Bool
 	workersDone      chan interface{}
 	workers          map[string]*worker = make(map[string]*worker)
+	watcherIsEnabled bool
 )
 
 func initWorkers(opt []workerOpt) error {
@@ -162,7 +163,10 @@ func (worker *worker) startNewWorkerThread() {
 			// if we end up here, the worker has not been up for backoff*2
 			// this is probably due to a syntax error or another fatal error
 			if failureCount >= maxWorkerConsecutiveFailures {
-				panic(fmt.Errorf("workers %q: too many consecutive failures", worker.fileName))
+				if !watcherIsEnabled {
+					panic(fmt.Errorf("workers %q: too many consecutive failures", worker.fileName))
+				}
+				logger.Warn("many consecutive worker failures", zap.String("worker", worker.fileName))
 			} else {
 				failureCount += 1
 			}
@@ -206,7 +210,7 @@ func restartWorkersOnFileChanges(workerOpts []workerOpt) error {
 	if len(directoriesToWatch) == 0 {
 		return nil
 	}
-
+	watcherIsEnabled = true
 	restartWorkers := func() {
 		restartWorkers(workerOpts)
 	}

--- a/worker.go
+++ b/worker.go
@@ -25,10 +25,10 @@ type worker struct {
 
 const maxWorkerErrorBackoff = 1 * time.Second
 const minWorkerErrorBackoff = 100 * time.Millisecond
-const maxWorkerConsecutiveFailures = 5
+const maxWorkerConsecutiveFailures = 6
 
 var (
-	isInitialBoot    bool = true
+	watcherIsEnabled bool
 	workersReadyWG   sync.WaitGroup
 	workerShutdownWG sync.WaitGroup
 	workersAreReady  atomic.Bool
@@ -55,7 +55,6 @@ func initWorkers(opt []workerOpt) error {
 
 	workersReadyWG.Wait()
 	workersAreReady.Store(true)
-	isInitialBoot = false
 
 	return nil
 }
@@ -164,7 +163,7 @@ func (worker *worker) startNewWorkerThread() {
 			// if we end up here, the worker has not been up for backoff*2
 			// this is probably due to a syntax error or another fatal error
 			if failureCount >= maxWorkerConsecutiveFailures {
-				if isInitialBoot {
+				if !watcherIsEnabled {
 					panic(fmt.Errorf("workers %q: too many consecutive failures", worker.fileName))
 				}
 				logger.Warn("many consecutive worker failures", zap.String("worker", worker.fileName), zap.Int("failures", failureCount))
@@ -197,6 +196,7 @@ func stopWorkers() {
 
 func drainWorkers() {
 	watcher.DrainWatcher()
+	watcherIsEnabled = false
 	stopWorkers()
 	workerShutdownWG.Wait()
 	workers = make(map[string]*worker)
@@ -207,7 +207,8 @@ func restartWorkersOnFileChanges(workerOpts []workerOpt) error {
 	for _, w := range workerOpts {
 		directoriesToWatch = append(directoriesToWatch, w.watch...)
 	}
-	if len(directoriesToWatch) == 0 {
+	watcherIsEnabled = len(directoriesToWatch) > 0
+	if !watcherIsEnabled {
 		return nil
 	}
 	restartWorkers := func() {

